7. In Sections 5.3.2 and 5.3.3, we saw that the cv.glm() function can be used in order to compute 
   the LOOCV test error estimate. Alternatively, one could compute those quantities using just 
   the glm() and predict.glm() functions, and a for loop. You will now take this approach in order 
   to compute the LOOCV error for a simple logistic regression model on the Weekly data set. 
   Recall that in the context of classification problems, the LOOCV error is given in (5.4).

**A.** Fit a logistic regressionmodel that predicts Direction using Lag1 and Lag2.

In [ ]:
require(ISLR)
summary(Weekly)

In [2]:
logreg <- glm(Direction~Lag1+Lag2,data = Weekly,family = "binomial")
summary(logreg)


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = "binomial", data = Weekly)

Deviance Residuals: 
   Min      1Q  Median      3Q     Max  
-1.623  -1.261   1.001   1.083   1.506  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22122    0.06147   3.599 0.000319 ***
Lag1        -0.03872    0.02622  -1.477 0.139672    
Lag2         0.06025    0.02655   2.270 0.023232 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1496.2  on 1088  degrees of freedom
Residual deviance: 1488.2  on 1086  degrees of freedom
AIC: 1494.2

Number of Fisher Scoring iterations: 4


**B.** Fit a logistic regression model that predicts Direction using Lag1
and Lag2 using all but the first observation.

In [12]:
logreg.sub <- glm(Direction~Lag1+Lag2,data = Weekly[-1,],family = "binomial")
summary(logreg.sub)


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = "binomial", data = Weekly[-1, 
    ])

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.6258  -1.2617   0.9999   1.0819   1.5071  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22324    0.06150   3.630 0.000283 ***
Lag1        -0.03843    0.02622  -1.466 0.142683    
Lag2         0.06085    0.02656   2.291 0.021971 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1494.6  on 1087  degrees of freedom
Residual deviance: 1486.5  on 1085  degrees of freedom
AIC: 1492.5

Number of Fisher Scoring iterations: 4


**C.** Use the model from (b) to predict the direction of the first observation. 
<br> You can do this by predicting that the first observation will go up if:
$P(Direction="Up"|Lag1, Lag2) > 0.5$. <br> Was this observation correctly classified?

In [17]:
logreg.prob <- predict(logreg.sub,newdata = Weekly[1,],type = "response")
logreg.pred <- ifelse(logreg.prob > 0.5,"Up","Down")
prediction <- ifelse(logreg.pred == Weekly[1,]$Direction,"Correct","misclassified")
prediction

[1] "misclassified"

Observation #1 was misclassified

**D.** Write a for loop from $i = 1$ to $i = n$, where $n$ is the number of observations in the data set, 
    that performs each of the following steps:

In [27]:
error <- rep(0,nrow(Weekly))
for (i in 1:nrow(Weekly)){
    # i. Fit a logistic regression model using all but the ith observation to predict Direction using Lag1 and Lag2.
    logreg.sub <- glm(Direction~Lag1+Lag2,data = Weekly[-i,],family = "binomial")
    # ii. Compute the posterior probability of the market moving up for the ith observation.
    logreg.prob <- predict(logreg.sub,newdata = Weekly[i,],type = "response") 
    # iii. Use the posterior probability for the ith observation in order to predict whether or not the market moves up.
    logreg.pred <- ifelse(logreg.prob > 0.5,"Up","Down")
    #iv. Determine whether or not an error was made in predicting the direction for the ith observation. If an error was made,
    #    then indicate this as a 1, and otherwise indicate it as a 0.
    error[i] <- ifelse(logreg.pred != Weekly[i,]$Direction,1,0)
}


**E.** Take the average of the n numbers obtained in (d)iv in order to obtain the LOOCV estimate for the test error. Comment on the
results.

In [28]:
mean(error)

[1] 0.4499541

The LOOCV error rate is 45%, which is significantly high (near random error). 
This result could be attributed to Lag1 and Lag2 not being good predictors of the Direction